### PR TITLE
Refactor - new map view

### PR DESCRIPTION
**Pull request by jankrzempek**
**What was done:**
In this PR `GoogleMapView.swift` was fully replaced by a new implementation, which will be temporarily named '_MapView`. 

`CreatingSessionMapView.swift` was deleted and replaced by the instance of `_NewMap`.
`SearchCompleteScreenMapView.swift` was deleted and replaced by the instance of `_NewMap`. 

**What was not changed and why:**
`SearchAndFollowMap` was not changed at all here and its purpose is to stay as it is. The reason is the following:
- this map has its specified purpose 
- was implemented in the near past, so the code seems to be better
- changing it would require more time

**_NewMap Description:**
The API for using the new map was developed, based on the official GoogleMaps docs and its API.

To use a new implementation just type `_NewMap(args)`
where args can be defined as shown below:

`path: [PathPoint] = []` - points used to draw a path.
`type: MapType` - can be normal (default) or satellite etc.
`trackingStyle: TrackingStyle` - it's all about the camera, and where it should be moved to. Do not consider user indicator here.
`userIndicatorStyle: UserIndicatorStyle` - it's all about the user's location, and his representation on the map.
`userTracker: UserTracker` - it is a protocol-based bedded implementation, so we can keep track of the user.
`markers: [Marker] = []` - it's all about additional things we would like to put on top of the map. In our case, this is used for notes.

### CODE REVIEW

File: AirCasting/CreateSessionViews/ChooseCustomLocationView.swift [hunk 1]
```diff
@@ -89,4 +95,17 @@ struct ChooseCustomLocationView: View {
             }
         )
     }
+    
+    private func updateTrackerWithNewLocation(_ loc: CLLocationCoordinate2D?) {
+        guard let loc else { return }
+        locationTracker.locationSource = loc
+        ApplocationTracker.location.value = .init(latitude: loc.latitude,
```
Comment (anna1901) [inline AirCasting/CreateSessionViews/ChooseCustomLocationView.swift:102]: Are you sure we want to update this AppLocationTracker location here? Shouldn't it always have actual location of the device? Why do we need to change it here?

**Review by anna1901 — CHANGES_REQUESTED** [review]
Wow, nice work, really! The new interface of the map is beautiful. 🤩
Issues I noticed:
- For mobile dormant session the zoom is way too big.
- When recording mobile session:
a) When I change zoom or move camera it is immediately moved back to default when new measurement comes
b) measurement dot color is changing between yellow and green when it should just stay green
c) heatmap color is changing to grey randomly
- When I finished an active session I got this log: "Stopped location tracking (refcount: 2)", and the refcount should actually be 0
- When I create a fixed session and I search for location eg. Warsaw, then my location button doesn't take me to my location anymore. Instead it takes me to the center of Warsaw.
- Additionally, the location chosen on custom location view is not shown on confirmation screen. Instead it shows my actual location. But the session location is correct (the one I chosen, not my actual location)

I haven't finished reviewing the code yet, but maybe you can start working on those issues already.

File: AirCasting/CreateSessionViews/ChooseCustomLocationView.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/CreateSessionViews/ChooseCustomLocationView.swift:102]: It generally should have the actual location, but as the implementations are different now, based on the use case, we need to handle it here the other way. In this case, we can get location updates from map and PlacePicker as well, so we decided it is an appropriate approach to update it from here.

File: AirCasting/CreateSessionViews/ChooseCustomLocationView.swift [hunk 2]
```diff
@@ -35,22 +37,26 @@ struct ChooseCustomLocationView: View {
...
 
     var mapGoogle: some View {
-        GoogleMapView(pathPoints: [],
-                      placePickerIsUpdating: $placePickerIsUpdating,
-                      isUserInteracting: Binding.constant(true),
-                      mapNotes: .constant([]),
-                      isMapOnPickerScreen: true,
-                      placePickerLocation: $location)
+        _MapView(type: .normal,
+                 trackingStyle: .user,
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/CreateSessionViews/ChooseCustomLocationView.swift:51]: Always the user decides where he is at the moment and provides that information to the app. He is in control and we believe in his information, so I think `user` is appropriate here. In the process of setting it up, we are not sure where he is, but when he is choosing it, we follow.

File: AirCasting/Map/_Map/_MapView.swift [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        
+        if isUserPositionTrackingRequired {
+            userTracker.startTrackingUserPosition { [weak coord = context.coordinator, weak map = mapView] location in
+                let oldValue = CLLocationCoordinate2D(latitude: location.coordinate.latitude, longitude: location.coordinate.longitude)
+                if (mapView.camera.target.latitude != oldValue.latitude) || (mapView.camera.target.longitude != oldValue.longitude) {
+                    coord?.latestUserLocation = location
+                    context.coordinator.suspendTracking = false
+                }
+                guard let map, let coord else { return }
+                setupUserIndicatorStyle(with: userIndicatorStyle, in: map, with: coord)
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/Map/_Map/_MapView.swift:123]: right, so that we can always update user representing `dot` when location changes

File: AirCasting/CreateSessionViews/ChooseCustomLocationView.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (anna1901) [reply AirCasting/CreateSessionViews/ChooseCustomLocationView.swift]: But what for? Isn't it why you introduced the BindableLocationTracker?

File: AirCasting/Map/AirMapView.swift [hunk 4]
```diff
@@ -66,15 +88,44 @@ struct AirMapView: View {
...
+                            // - no need for real time updated
+                            _MapView(path: pathPoints,
+                                     type: .normal,
+                                     trackingStyle: .wholePath,
+                                     userIndicatorStyle: .none,
+                                     locationTracker: MapLocationTrackerAdapter(locationTracker),
+                                     markers: mapNotesVM.notes.asMapMarkers(with: didTapNote))
+                            .addingOverlay { mapView in overlayHeatMap(on: mapView, threshold: threshold) }
+                        }
+                        #warning("TODO: Implement calculating stats only for visible path points")
```
Comment (anna1901) [inline AirCasting/Map/AirMapView.swift:124]: This can be removed. Michael said it is not needed anymore (and I think this comment is not present on develop branch anymore either)

File: AirCasting/Map/_Map/_MapView.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (anna1901) [reply AirCasting/Map/_Map/_MapView.swift:123]: So I think the name should be changed to `setupLocationIndicator` or something like that.

File: AirCasting/Map/_Map/_MapView.swift [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+    
+    func drawPolyline(_ uiView: GMSMapView, coordinator: Coordinator) {
+        let newPath = GMSMutablePath()
+        for point in path {
+            newPath.add(.init(latitude: point.lat, longitude: point.long))
+        }
+        
+        let polyline = coordinator.polyline
+        polyline.path = newPath
+        polyline.strokeColor = .accentColor // TODO: Change it (styling)
```
Comment (anna1901) [inline AirCasting/Map/_Map/_MapView.swift:318]: When are you gonna change those TODOs?

File: AirCasting/Map/_Map/_MapView.swift [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+        polyline.strokeWidth = CGFloat(3)
+        polyline.map = uiView
+    }
+    
+    private func followUser(in mapView: GMSMapView, coordinator: Coordinator) {
+        mapView.isMyLocationEnabled = true
+        // TODO: Check on device if we can use the `LocationTracker` instead
```
Comment (anna1901) [inline AirCasting/Map/_Map/_MapView.swift:325]: And this one

File: AirCasting/Settings/ShareLogsViewModel.swift [hunk 7]
```diff
@@ -11,7 +11,7 @@ protocol LogfileProvider {
 class ShareLogsViewModel: ObservableObject {
     @Published var shareSheetPresented: Bool = false
     var file: URL? { logFileProvider.logFileURLForSharing() }
-    @Injected private var logFileProvider: LogfileProvider
+    private var logFileProvider: LogfileProvider { LoggerBuilder.shared.store }
```
Comment (anna1901) [inline AirCasting/Settings/ShareLogsViewModel.swift:14]: It would be awesome if those changes related to Logger could to be moved to another PR. Do you think you could do this? I think it would be very beneficial for keeping track of changes made in the project when looking at the develop branch (where each PR is squashed to one commit).

File: AirCasting/Map/AirMapView.swift [hunk 8]
```diff
@@ -11,6 +11,12 @@ import Foundation
 import CoreData
 import Resolver
 
+class HeatmapContainer: ObservableObject {
+    // It keeps track of heatmap instances
+    // so that we always have only one
+    var heatMap: Heatmap?
+}
+
 struct AirMapView: View {
```
[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/Map/AirMapView.swift:20]: You right, I don't know the meaning but sounds fancy for sure. I like it 😅

[the remark this answers is not in the document]
Comment (jankrzempek) [reply AirCasting/Map/AirMapView.swift:20]: I will change ofc
